### PR TITLE
Formatter

### PR DESCRIPTION
**Pull request by alexeagle**
The gulp task is working (warns if your file isn't formatted).
But the resulting format is still buggy, eg. \n in the middle of a => operator.

### CODE REVIEW

**Comment by alexeagle** [comment]
Also doesn't understand backtick'ed string literals.

**Comment by mprobst** [comment]
There still must be something funky going on with the version of clang-format that you're running. Backtick literals should work (or at least survive untouched), I have tests to prove it: https://github.com/llvm-mirror/clang/blob/master/unittests/Format/FormatTestJS.cpp#L577

Same for fat arrows. Are you sure you're installing/compiling the HEAD version?

**Comment by alexeagle** [comment]
Yeah, version skew. There is a great deal of confusion in the many places node_modules have copied the binary, the places homebrew has put it, and I get many different behaviors, so that's my excuse.

Updated this PR to have the right formatted result.

File: main.ts [hunk 1]
```diff
@@ -48,7 +46,8 @@ export function translateProgram(program: ts.Program): string {
     throw new Error(`${file.fileName}:${pos.line}:${pos.character}: ${message}`);
   }
 
-  // Comments attach to all following AST nodes before the next 'physical' token. Track the earliest
+  // Comments attach to all following AST nodes before the next 'physical'
+  // token. Track the earliest
```
Comment (mprobst) [inline main.ts]: Please re-indent (clang-format doesn't format comments).

File: main.ts [hunk 2]
```diff
@@ -160,10 +159,11 @@ export function translateProgram(program: ts.Program): string {
         visit(forStmt.statement);
         break;
       case ts.SyntaxKind.ForInStatement:
-        // TODO(martinprobst): Dart's for-in loops actually have different semantics, they are more
+        // TODO(martinprobst): Dart's for-in loops actually have different
```
Comment (mprobst) [inline main.ts]: Same re comment.

File: main.ts [hunk 3]
```diff
@@ -192,11 +192,11 @@ export function translateProgram(program: ts.Program): string {
 
       // Literals.
       case ts.SyntaxKind.NumericLiteral:
-        var sLit = <ts.LiteralExpression>node;
+        var sLit = <ts.LiteralExpression> node;
         emit(sLit.getText());
         break;
       case ts.SyntaxKind.StringLiteral:
-        var sLit = <ts.LiteralExpression>node;
+        var sLit = <ts.LiteralExpression> node;
```
Comment (mprobst) [inline main.ts]: Might be worth fixing this, I like the no-whitespace version better (though OTOH it has lower precedence, <X> foo.bar is <X> (foo.bar), isn't it?).

File: main.ts [hunk 4]
```diff
@@ -387,11 +387,14 @@ export function translateProgram(program: ts.Program): string {
...
+    module: ts.ModuleKind.CommonJS
+  };
   var host = ts.createCompilerHost(options);
   var program = ts.createProgram(fileNames, options, host);
   return translateProgram(program);
 }
 
 // CLI entry point
-translateFiles(process.argv.slice(2));
+translateFiles(process.argv.slice(2));
```
Comment (mprobst) [inline main.ts]: Did it actually eat the newline at the end of the file? It shouldn't.

Comment (alexeagle) [inline main.ts]: Yeah I believe it did eat the final newline. About space after cast, we
should also check what is typical for TS code... I think it is no space.

File: main.ts (hunk 1 above)
Comment (alexeagle) [inline main.ts]: doesn't seem to be an issue since I'm now allowing 100c

File: main.ts (hunk 4 above)
Comment (alexeagle) [inline main.ts]: I was wrong, I ran the formatter again and it didn't eat a final newline. Looks pretty good, will commit now.
Next step is to format the test and then turn on the enforcement somehow.

**Pull request merged**